packages/transaction-messages/README.md

```diff
@@ -14,4 +14,213 @@
+
+### Types
+
+#### `TransactionMessageWithBlockhashLifetime`
+
+This type represents a transaction message whose lifetime is defined by the age of the blockhash it includes. Such a transaction can only be landed on the network if the current block height of the network is less than or equal to the value of `TransactionMessageWithBlockhashLifetime['lifetimeConstraint']['lastValidBlockHeight']`.
+
+#### `TransactionMessageWithDurableNonceLifetime`
+
+This type represents a transaction message whose lifetime is defined by the value of a nonce it includes. Such a transaction can only be landed on the network if the nonce is known to the network and has not already been used to land a different transaction.
+
+#### `Blockhash`
+
+This type represents a string that is particularly known to be the base58-encoded value of a block.
+
+#### `Nonce`
+
+This type represents a string that is particularly known to be the base58-encoded value of a nonce.
+
+### Functions
+
+#### `setTransactionMessageLifetimeUsingBlockhash()`
+
+Given a blockhash and the last block height at which that blockhash is considered usable to land transactions, this method will return a new transaction message having the same type as the one supplied plus the `TransactionMessageWithBlockhashLifetime` type.
+
+```ts
+import { setTransactionMessageLifetimeUsingBlockhash } from '@solana/transaction-messages';
+
+const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
+const txWithBlockhashLifetime = setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx);
+```
+
+#### `setTransactionMessageLifetimeUsingDurableNonce()`
+
+Given a nonce, the account where the value of the nonce is stored, and the address of the account authorized to consume that nonce, this method will return a new transaction having the same type as the one supplied plus the `TransactionMessageWithDurableNonceLifetime` type. In particular, this method _prepends_ an instruction to the transaction message designed to consume (or &lsquo;advance&rsquo;) the nonce in the same transaction whose lifetime is defined by it.
+
+```ts
+import { setTransactionMessageLifetimeUsingDurableNonce } from '@solana/transactions';
+
+const NONCE_VALUE_OFFSET =
+    4 + // version(u32)
+    4 + // state(u32)
+    32; // nonce authority(pubkey)
+// Then comes the nonce value.
+
+const nonceAccountAddress = address('EGtMh4yvXswwHhwVhyPxGrVV2TkLTgUqGodbATEPvojZ');
+const nonceAuthorityAddress = address('4KD1Rdrd89NG7XbzW3xsX9Aqnx2EExJvExiNme6g9iAT');
+const { value: nonceAccount } = await rpc
+    .getAccountInfo(nonceAccountAddress, {
+        dataSlice: { length: 32, offset: NONCE_VALUE_OFFSET },
+        encoding: 'base58',
+    })
+    .send();
+const nonce =
+    // This works because we asked for the exact slice of data representing the nonce
+    // value, and furthermore asked for it in `base58` encoding.
+    nonceAccount!.data[0] as unknown as Nonce;
```

Comment:
Should we replace this, now, with:

```ts
import { fetchNonce } from '@solana-program/system';

const { data: { blockhash: nonce } } = await fetchNonce(rpc, nonceAccountAddress);
```

The existing version is way more efficient because of the base58 fetch trick, so I'm torn.

cc/ @lorisleiva

Reply:
I'm mostly concerned about the cyclic dependencies this would create. I think this is okay as-is for now.

Reply:
Gah! Again. Damnit.

Reply:
I think the best thing to do is to have a small Kinobi tree of any instruction/account we want to export on the main library and then generate them straight inside a package on the monorepo. That way everything is versioned together.

Reply:
FWIW this is just the readme - we wouldn't be adding the code/dependency into the `transaction-messages` package, just saying that if you want to do this then this is how to do it. There wouldn't be any issue with someone importing `@solana/transaction-messages` and `@solana-program/system` into the same code right? Assuming web3js versions are aligned of course.

Reply:
Ah yeah, my bad I wasn't thinking about it from the README's perspective haha. Yeah 100%, there is no issue replacing the README with the generated instruction. I assumed we were talking about incorporating generated clients such as `@solana-program/system` inside the main library.

Reply:
I'm clearly not thinking at all.